### PR TITLE
4.0.7.8

### PR DESCRIPTION
**Pull request by Ghost-chu**
1. Optimized Cache logic.
2. Only use a single item and center it for a Double-Chest shop ( thanks to @Starmism ).
3. Add Adventure GameMode support.
4. Add swap behavior between right and left click.
5. Add custom amount for command `/qs create`.
6. Removed ArmorStand display.
7. Optimized VDI costed too many CPU time.

### CODE REVIEW

**Review by Ghost-chu — COMMENTED** [review]
deleted

File: src/main/java/org/maxgamer/quickshop/QuickShop.java [hunk 1]
```diff
@@ -808,7 +808,7 @@ public void onEnable() {
         Bukkit.getPluginManager().registerEvents(internalListener, this);
         if (isDisplay() && DisplayItem.getNowUsing() != DisplayType.VIRTUALITEM) {
             displayWatcher = new DisplayWatcher(this);
-            new DisplayBugFixListener(this).register();
+//            new DisplayBugFixListener(this).register();
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/QuickShop.java:811]: Why disable this?

File: src/main/java/org/maxgamer/quickshop/integration/superiorskyblock/SuperiorSkyblock2Integration.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+public class SuperiorSkyblock2Integration extends QSIntegratedPlugin implements Listener {
+    private final boolean onlyOwnerCanCreateShop;
+    private final boolean deleteShopOnMemberLeave;
+
+    public SuperiorSkyblock2Integration(QuickShop plugin) {
+        super(plugin);
+        onlyOwnerCanCreateShop = plugin.getConfig().getBoolean("integration.superiorskyblock.owner-create-only");
+        deleteShopOnMemberLeave = plugin.getConfig().getBoolean("integration.superiorskyblock.delete-shop-on-member-leave");
+        if (plugin.getConfig().getBoolean("integration.superiorskyblock.delete-shop-on-member-leave")) {
+            Bukkit.getPluginManager().registerEvents(this, plugin);
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/integration/superiorskyblock/SuperiorSkyblock2Integration.java:49]: Please put register and unregister event invoke to `load` and `unload` method

File: src/main/resources/config.yml [hunk 3]
```diff
@@ -663,6 +680,13 @@ integration:
         warzone: false
         #Player must have flags in the faction:
       flags: [ ]
+  #Superior Skyblock Integration
+  superiorskyblock:
+    enable: false
+    #Does only island owner can create shops on island?
+    owner-create-only: false
+    #Delete shop when the shop owner kicked/quited from island
+    delete-shop-on-member-leave: true
```
Comment (sandtechnology) [inline src/main/resources/config.yml:689]: Please add update code in config

File: src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java [hunk 4]
```diff
@@ -370,15 +371,26 @@ public void onPacketSending(@NotNull PacketEvent event) {
...
+//        }, 0, 1);
+    }
+
+    private void unload() {
+        packetSenders.clear();
+        if (packetAdapter != null) {
+            protocolManager.removePacketListener(packetAdapter);
+        }
+        if (asyncSendingTask != null && !asyncSendingTask.isCancelled()) {
+            asyncSendingTask.cancel();
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/VirtualDisplayItem.java:389]: It is not needed since `asyncSendingTask` is always null.

**Review by sandtechnology — CHANGES_REQUESTED** [review]
Read above comments, and I'm not sure whether `Jsonconfiguration` is working

File: src/main/resources/config.yml (hunk 3 above)
Comment (Ghost-chu) [reply src/main/resources/config.yml:689]: L1727-L1730

File: src/main/java/org/maxgamer/quickshop/QuickShop.java (hunk 1 above)
Comment (Ghost-chu) [reply src/main/java/org/maxgamer/quickshop/QuickShop.java:811]: It designed for ArmorStandDisplay and it now deleted

Comment (Ghost-chu) [reply src/main/java/org/maxgamer/quickshop/QuickShop.java:811]: It designed for ArmorStandDisplay and it now deleted

**Review by sandtechnology — APPROVED** [review]

**Pull request merged**